src/Java.Interop.Tools.JavaTypeSystem/Adapters/ManagedApiImporter.cs

```diff
@@ -0,0 +1,379 @@
+using System.Linq;
+using Java.Interop.Tools.Cecil;
+using Java.Interop.Tools.JavaTypeSystem.Models;
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Java.Interop.Tools.JavaTypeSystem
+{
+	public static class ManagedApiImporter
+	{
+		public static JavaTypeCollection Parse (AssemblyDefinition assembly, JavaTypeCollection collection)
+		{
+			var types_to_add = new List<JavaTypeModel> ();
+
+			foreach (var md in assembly.Modules)
+				foreach (var td in md.Types) {
+					if (!ShouldSkipType (td) && ParseType (td, collection) is JavaTypeModel type)
+						types_to_add.Add (type);
+				}
+
+			// This needs to be done ordered from least nested to most nested, in order for nesting to work.
+			// That is, 'android.foo.Blah' needs to be added before 'android.foo.Blah.Bar'.
+			// Plus, we may have unnested managed types that are actually nested in Java-land:
+			// ex: IContextMenu and IContextMenuContextMenuItem
+			foreach (var type in types_to_add.OrderBy (t => t.FullName.Count (c => c == '.')).ToArray ())
+				AddReferenceTypeRecursive (type, collection);
+
+			return collection;
+		}
+
+		public static JavaTypeModel? ParseType (TypeDefinition type, JavaTypeCollection collection)
+		{
+			if (!type.IsPublic && !type.IsNested)
+				return null;
+
+			if (!ShouldImport (type))
+				return null;
+
+			var model = type.IsInterface ? (JavaTypeModel?) ParseInterface (type, collection) : ParseClass (type, collection);
+
+			if (model is null)
+				return null;
+
+			foreach (var nested in type.NestedTypes)
+				if (ParseType (nested, collection) is JavaTypeModel nested_model)
+					model.NestedTypes.Add (nested_model);
+
+			return model;
+		}
+
+		static bool ShouldImport (TypeDefinition td)
+		{
+			// We want to exclude "IBlahInvoker" and "IBlahImplementor" and "BlahConsts" types
+			if (td.Name.EndsWith ("Invoker")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 7);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			if (td.Name.EndsWith ("Implementor")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 11);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			if (td.Name.EndsWith ("Consts")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 6);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			return true;
+		}
+
+		public static JavaClassModel? ParseClass (TypeDefinition type, JavaTypeCollection collection)
+		{
+			// TODO: type parameters?
+			var obs_attr = GetObsoleteAttribute (type.CustomAttributes);
+			var reg_attr = GetRegisterAttribute (type.CustomAttributes);
+
+			if (reg_attr is null)
```

Comment:
The implication here is that `JavaTypeSystem` *only* supports bindings, and not "Java interop".  For example, if I create a new class:

```csharp
class Example : Java.Lang.Object {
}
```

then `Example` *will not* be parsed, as it lacks a `[Register]`.  Is this desirable?

Why not instead/alternatively use [`JavaNativeTypeManager.GetJniTypeName()`](https://github.com/xamarin/java.interop/blob/855ecfa3a85ce74f40a1cc47e67b70aedd697dc3/src/Java.Interop.Tools.TypeNameMappings/Java.Interop.Tools.TypeNameMappings/JavaNativeTypeManager.cs#L458), which would support `Example`?

Reply:
I think the implication is that `JavaTypeSystem` is only for supporting `generator` usage.  I do not think it is a generic enough library to support other use cases.